File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;

public class Main {
	static int N, count;
	static final int[] DIRX = {1, 0, -1, 0}; //상 우 하 좌
	static final int[] DIRY = {0, 1, 0, -1};
	static int[][] map;
	
	static PriorityQueue<Integer> result;
	static ArrayDeque<int[]> ad;
		
	static void bfs(int x, int y) {
		ad.offer(new int[] {x, y});
		map[x][y] = 0;
		count++;
		
		int check = 0;
		while (!ad.isEmpty()) {
			check++;
			
			int[] now = ad.poll();
			
			for (int d = 0; d < 4; d++) {
				int dx = now[0] + DIRX[d];
				int dy = now[1] + DIRY[d];
				
				if (dx >= 0 && dx < N && dy >= 0 && dy < N && map[dx][dy] == 1) {
					map[dx][dy] = 0;
					ad.offer(new int[] {dx, dy});
				}
			}
		}
		result.offer(check);
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		count = 0;
		
		map = new int[N][N];
		ad = new ArrayDeque<>();
		result = new PriorityQueue<>(new Comparator<Integer>() {
			@Override
			public int compare(Integer o1, Integer o2) {
				if (o1 > o2) {
					return 1;
				} else {
					return -1;
				}
			}
		});
		
		for (int i = 0; i < N; i++) {
			char[] ch = br.readLine().toCharArray();
			for (int j = 0; j < N; j++) {
				map[i][j] = Character.valueOf(ch[j]) - '0';
			}
		}
		
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if(map[i][j] == 1) {
					bfs(i, j);
				}
			}
		}
		sb.append(count + "\n");
		
		while (!result.isEmpty()) {
			sb.append(result.poll() + "\n");
		}
		System.out.println(sb);
		br.close();
	}
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
dirX = [-1,1,0,0]
dirY = [0,0,-1,1]

import sys
input = sys.stdin.readline

N = int(input())

town = []

for i in range(N) :
    town.append(list(map(int, str(input().strip()))))

result = []

def bfs(x, y) :
    pos = []
    pos.append([x, y])
    town[x][y] = 0
    compare = 1

    while len(pos) > 0 :
        pop = pos.pop()

        for i in range(4) :
            dx = pop[0] + dirX[i]
            dy = pop[1] + dirY[i]

            if 0<=dx and dx<N and 0<=dy and dy<N :
                if town[dx][dy] == 1 :
                    town[dx][dy] = 0
                    compare += 1
                    pos.append([dx, dy])

    return compare

for i in range(N) :
    for j in range(N) :
        if town[i][j] == 1 :
            result.append(bfs(i, j))

result.sort()

size = len(result)
print(size)
for i in range(size) :
    print(result[i])